### PR TITLE
continuation of conversion of multihash to string

### PR DESCRIPTION
**Pull request by kevina**
Continuation of #83 using parts of #82.

### CODE REVIEW

File: multihash.go [hunk 1]
```diff
@@ -227,6 +229,36 @@ func FromB58String(s string) (m Multihash, err error) {
 	return Cast(b)
 }
 
+// FromBinary creates a new multihash from the binary
+// representation. The string is assumed to be a valid multihash.
+// This function will not alloc and will panic if it is unable to
+// parse the string.
+func FromBinary(v string) Multihash {
```
Comment (Stebalien) [inline multihash.go]: Any reason not to have this return an error? I don't see this being called *so* often checking an error becomes onerous. For example, I expect we'll want to use this from go-cid when validating the CID's multihash.

(It may *also* be useful to have a `FromBinaryUnchecked` that doesn't even check but that's a separate issue).

File: multihash.go [hunk 2]
```diff
@@ -227,6 +229,36 @@ func FromB58String(s string) (m Multihash, err error) {
...
+	i := strbinary.UvarintLen(v)
+	digestLen, l := strbinary.Uvarint(v[i:])
+	i += l
+	if len(v[i:]) != int(digestLen) {
+		panic(ErrInvalidMultihash)
+	}
+	return Multihash{v}
+}
+
+// Parts returns the components of the Multihash (Code, Name, Digest)
```
[the remark this answers is not in the document]
Comment (kevina) [reply multihash.go]: Yeah your right since the length is primary useful in the binary representation as a sanity check and to know when the multihash ends in a byte stream.

File: multihash.go (hunk 1 above)
Comment (kevina) [reply multihash.go]: I wanted something that was fast and convenient to use when we _know_ the multihash is already valid.  The error checking that is done are sanity checks.  For example if the hash length is incorrect or bogus this will likely panic due to an out of bounds index check rather than return a proper error message.

I am somewhat against FromBinaryUnchecked if we can get away with (performance wise) always doing the checks.  This way we _know_ that the Multihash is valid if it is not empty.

See https://github.com/ipfs/go-cid/pull/73 for how I use this.  I am less against fixing this to return an error.

Comment (Stebalien) [reply multihash.go]: Got it. I'd like to expose *some* `FromBinary` function that returns an error. Currently, `go-cid` will have to allocate to check the multihash. That is, it'll:

1. Pass a `[]byte` into `mh.Cast`.
2. `mh.Cast` will return a `string` backed multihash (allocating).

A `FromBinary` that returns an error would allow us to check without panicing or allocating.

Comment (kevina) [reply multihash.go]: Okay.  I created such a function and called it `New`.

Comment (Stebalien) [reply multihash.go]: Why `New`? It doesn't really create a *new* multihash.

> I am somewhat against FromBinaryUnchecked if we can get away with (performance wise) always doing the checks. This way we know that the Multihash is valid if it is not empty.

But we *aren't* always doing the checks! `FromString` leaves out one of the checks. Let's either check everything or nothing, checking somethings buys us nothing and just adds a bunch of code and two confusing functions that do slightly different things.

Comment (kevina) [reply multihash.go]: Okay fair enough.  I renamed New to FromBinary when the Cid needs to return a Multihash it can just ignore the errors.

If necessary e can add an unchecked version if this becomes a performance problem (which I doubt).

Comment (Stebalien) [reply multihash.go]: As you said, we can consider adding an unchecked variant later. My point was that we can actually put "checks" in the unchecked version (as long as they're fast enough) as long as we don't make any promises. My issue with the previous `FromBinary` incarnation is that it sort of checked and made unclear guarantees. I'd rather have a fully checked version (that returns the error) and a version that makes absolutely no guarantees (but *can* choose to check under the covers).

TL;DR: I'm happy with the current version, we can take a look into a less checked variant later.

**Comment by kevina** [comment]
@Stebalien  Do you want to get this is now with the Cid changes?

**Comment by kevina** [comment]
@Stebalien I can try (the full gx-update with multihash).  It may be helpful if I do it at a time when you are around so you can walk me though and make sure everything looks good.

So that I am confident you are okay with all the changes please Approve this p.r., https://github.com/ipfs/go-cid/pull/71 and https://github.com/ipfs/go-cid/pull/73.

**Comment by Stebalien** [comment]
Only @kevina should merge this. CI fails because this breaks go-libp2p-peer.

@kevina modulo the one comment on go-cid, this all looks good to me. I'd:

1. Start by *not* using gx. Instead, just fix all the appropriate packages (picking up where I left off), testing with go.
2. Once that all works, we can use gx-workspace to make this happen (if it looks too gnarly, I can take a stab at this; this'll be interesting...).

**Comment by lidel** [comment]
Closing due to being stale since 2018.
